### PR TITLE
Add Gzip request compression feature

### PR DESCRIPTION
**Pull request by wty-Bryant**
Add Gzip request compression feature to operations supporting that trait. Insert middleware to those op's stack and enable toggling the compression via client options, env config and shared config profile.

### CODE REVIEW

File: .changelog/a96149e35456411aaca3224016feff3c.json [hunk 1]
```diff
@@ -0,0 +1,16 @@
+{
+    "id": "a96149e3-5456-411a-aca3-224016feff3c",
+    "type": "feature",
+    "collapse": true,
+    "description": "Add and Merge request compression feature to ops supporting that trait",
```
Comment (lucix-aws) [inline .changelog/a96149e35456411aaca3224016feff3c.json:5]:
```suggestion
    "description": "Support modeled request compression. The only algorithm supported at this time is `gzip`.",
```

**Review by wty-Bryant — COMMENTED** [review]
for this [issue](https://github.com/aws/aws-sdk-go-v2/pull/2344#issuecomment-1828472163), I actually generate and test `cloudwatch.PutMetricData()` op with the compression trait, I can publish that first for your testing

File: config/shared_config.go [hunk 2]
```diff
@@ -1099,6 +1118,56 @@ func (c *SharedConfig) setFromIniSection(profile string, section ini.Section) er
...
+	}
+	if v < 0 || v > 10485760 {
+		return fmt.Errorf("invalid range for min request compression size bytes %d, must be within 0 and 10485760 inclusively", v)
+	}
+	*bytes = new(int64)
+	**bytes = v
+	return nil
+}
+
+func updateDisableRequestCompression(disable **bool, sec ini.Section, key string) error {
```
[the remark this answers is not in the document]
Comment (wty-Bryant) [reply config/shared_config.go:1138]: Same reason above

File: config/env_config.go [hunk 3]
```diff
@@ -370,6 +389,41 @@ func (c EnvConfig) getAppID(context.Context) (string, bool, error) {
 	return c.AppID, len(c.AppID) > 0, nil
 }
 
+func setRequestMinCompressSizeBytes(bytes **int64) error {
```
[the remark this answers is not in the document]
Comment (lucix-aws) [reply config/env_config.go:405]: I would have just used/implemented `setInt[64]PtrFromEnv` and then inlined the range check, but this is fine.

File: config/env_config.go [hunk 4]
```diff
@@ -383,6 +402,41 @@ func (c EnvConfig) getAppID(context.Context) (string, bool, error) {
...
+	b := os.Getenv(awsRequestMinCompressionSizeBytes)
+	if b == "" {
+		return nil
+	}
+
+	byte, err := strconv.ParseInt(b, 10, 64)
+	if err != nil {
+		return fmt.Errorf("invalid value for env var, %s=%s, need int64",
+			awsRequestMinCompressionSizeBytes, b)
+	} else if byte < 0 || byte > 10485760 {
```
Comment (lucix-aws) [inline config/env_config.go:415]: can you pull (exporting if needed) the `const` from the smithy-go side?

File: config/shared_config.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (lucix-aws) [reply config/shared_config.go:1158]: FYI you can just use `updateBoolPtrFromEnv` or whatever it's called.

**Review by lucix-aws — APPROVED** [review]
fixandship, great work

File: config/shared_config.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (wty-Bryant) [reply config/shared_config.go:1158]: `updateBoolPtr` in sharedconfig doesn't return error for invalid value, I will just keep that

**Pull request merged**